From 7dc5f84e8b4fb27a67b6a6790379492383bbc64f Mon Sep 17 00:00:00 2001
Message-Id: <7dc5f84e8b4fb27a67b6a6790379492383bbc64f.1423872525.git.feitong.yi@intel.com>
In-Reply-To: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
References: <3062c7e6b919c393846ff0e13558e7f648b352ca.1423872525.git.feitong.yi@intel.com>
From: Gaurav K Singh <gaurav.k.singh@intel.com>
Date: Tue, 20 Jan 2015 13:57:38 +0530
Subject: [PATCH 60/95] MUST_REBASE [VPG]: drivers/video/adf: Fix for MIPI
 dual display

This patch has the following fixes:
    1. Disable DPOunit clock gating
    2. Enable DPLL A ref clock bit in the PLL register

In single display, GOP takes care of programming the above mentioned registers.
But when HDMI & MIPI are connected in dual display mode, GOP programs only HDMI.
For such scenarios, MIPI driver must program these registers.

MUST_REBASE: Since Google ADF framework is not in upstream and also
we cannot have two display drivers, other one being i915, we need to
work on i915 and ADF convergence path before it can be upstreamed

For: GMINL-5348
Change-Id: I05b9aa482a217fbc0606a96150d4e7103d159289
Signed-off-by: Gaurav K Singh <gaurav.k.singh@intel.com>
---
 drivers/video/adf/intel/core/vlv/vlv_dsi_pll.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/video/adf/intel/core/vlv/vlv_dsi_pll.c b/drivers/video/adf/intel/core/vlv/vlv_dsi_pll.c
index 0cb749b..d6fd032 100644
--- a/drivers/video/adf/intel/core/vlv/vlv_dsi_pll.c
+++ b/drivers/video/adf/intel/core/vlv/vlv_dsi_pll.c
@@ -253,6 +253,7 @@ u32 vlv_dsi_pll_enable(struct vlv_pll *pll,
 		struct drm_mode_modeinfo *mode)
 {
 	u32 val;
+	u32 temp = 0;
 
 	/* Disable DPOunit clock gating, can stall pipe */
 	val = REG_READ(pll->offset);
@@ -261,6 +262,7 @@ u32 vlv_dsi_pll_enable(struct vlv_pll *pll,
 
 	val = REG_READ(DSPCLK_GATE_D);
 	val |= VSUNIT_CLOCK_GATE_DISABLE;
+	val |= DPOUNIT_CLOCK_GATE_DISABLE;
 	REG_WRITE(DSPCLK_GATE_D, val);
 
 	vlv_dsi_pll_configure(pll);
@@ -270,6 +272,12 @@ u32 vlv_dsi_pll_enable(struct vlv_pll *pll,
 
 	val = vlv_cck_read(CCK_REG_DSI_PLL_CONTROL);
 	val |= DSI_PLL_VCO_EN;
+
+	temp = REG_READ(pll->offset);
+	temp |= DPLL_REFA_CLK_ENABLE_VLV;
+	REG_WRITE(pll->offset, temp);
+	udelay(1000);
+
 	vlv_cck_write(CCK_REG_DSI_PLL_CONTROL, val);
 
 	pr_info("DSI PLL locked\n");
-- 
1.7.9.5

